### PR TITLE
Check nSpendsSapling, nOutputsSapling, and nActionsOrchard 2^16 limit

### LINKED ISSUE #2379: Do we need to validate nSpendsSapling, nOutputsSapling, and nActionsOrchard limits in V5 transactions?
## Motivation

We need to validate Orchard transactions from NU5 onwards.

Orchard signatures and proofs are already covered by other tickets, but we should check all the fields are being validated.

### Specifications

_TODO: Are there any other orchard or NU5 consensus rules?_

#### Shielded Input and Output Limits

> [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.

Explanation:
> [NU5 onward] The rule that nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16, is technically redundant because a transaction that could violate this rule would not fit within the 2 MB block size limit. It is included in order to simplify the security argument for balance preservation.

Note:
- We could implement this rule directly, but we already limit these sizes using `T::max_allocation` for spends, outputs, and actions. So we could just add a comment about this rule.
- We could also skip this rule, but add a compile-time assertion that it is redundant, using the same minimum spend, output, and action sizes as `T::max_allocation`

https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus

## Solution

_TODO: Work out how to check each consensus rule_

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

The spec says

> [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.

> [NU5 onward] The rule that nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16, is technically redundant because a transaction that could violate this rule would not fit within the 2 MB block size limit. It is included in order to simplify the security argument for balance preservation.

While the check is redudant, it seems useful to add a sanity check to make sure it's being enforced.

## Solution

Add static assertions to the current limits (which are derived from the block size) to make sure they are not greater than 2^16.

Closes https://github.com/ZcashFoundation/zebra/issues/2379

## Review

Anyone can review. Not urgent.

Closes https://github.com/ZcashFoundation/zebra/issues/2379

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
The macOS failure is unrelated, PR #3072 should fix it.

File: zebra-chain/src/orchard/shielded_data.rs [hunk 1]
```diff
@@ -182,7 +182,14 @@ impl TrustedPreallocate for Action {
...
         // and the signature is required,
         // a valid max allocation can never exceed this size
-        (MAX_BLOCK_BYTES - 1) / AUTHORIZED_ACTION_SIZE
+        const MAX: u64 = (MAX_BLOCK_BYTES - 1) / AUTHORIZED_ACTION_SIZE;
+        // > [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.
+        // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+        // This acts as nActionsOrchard and is therefore subject to the rule.
+        // The maximum value is actually smaller due to the block size limit,
+        // but we ensure the 2^16 limit with a static assertion.
+        static_assertions::const_assert!(MAX <= (1 << 16));
```
Comment (jvff) [inline zebra-chain/src/orchard/shielded_data.rs]: Should this be just less than?

```suggestion
        static_assertions::const_assert!(MAX < (1 << 16));
```

File: zebra-chain/src/sapling/output.rs [hunk 2]
```diff
@@ -217,7 +217,15 @@ impl TrustedPreallocate for OutputInTransactionV4 {
...
         // the max allocation can never exceed (MAX_BLOCK_BYTES - 1) / OUTPUT_SIZE
-        (MAX_BLOCK_BYTES - 1) / OUTPUT_SIZE
+        const MAX: u64 = (MAX_BLOCK_BYTES - 1) / OUTPUT_SIZE;
+        // > [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.
+        // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+        // This acts as nOutputsSapling and is therefore subject to the rule.
+        // The maximum value is actually smaller due to the block size limit,
+        // but we ensure the 2^16 limit with a static assertion.
+        // (The check is not required pre-NU5, but it doesn't cause problems.)
+        static_assertions::const_assert!(MAX <= (1 << 16));
```
Comment (jvff) [inline zebra-chain/src/sapling/output.rs]: Same here:

```suggestion
        static_assertions::const_assert!(MAX < (1 << 16));
```

File: zebra-chain/src/sapling/spend.rs [hunk 3]
```diff
@@ -276,7 +276,15 @@ impl TrustedPreallocate for SpendPrefixInTransactionV5 {
...
         // a valid max allocation can never exceed this size
-        (MAX_BLOCK_BYTES - 1) / SHARED_ANCHOR_SPEND_SIZE
+        const MAX: u64 = (MAX_BLOCK_BYTES - 1) / SHARED_ANCHOR_SPEND_SIZE;
+        // > [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.
+        // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+        // This acts as nSpendsSapling and is therefore subject to the rule.
+        // The maximum value is actually smaller due to the block size limit,
+        // but we ensure the 2^16 limit with a static assertion.
+        // (The check is not required pre-NU5, but it doesn't cause problems.)
+        static_assertions::const_assert!(MAX <= (1 << 16));
```
Comment (jvff) [inline zebra-chain/src/sapling/spend.rs]: Same here:

```suggestion
        static_assertions::const_assert!(MAX < (1 << 16));
```

File: zebra-chain/src/lib.rs [hunk 4]
```diff
@@ -22,6 +22,9 @@ extern crate serde_big_array;
 #[macro_use]
 extern crate bitflags;
 
+#[macro_use]
+extern crate static_assertions;
+
```
Comment (jvff) [inline zebra-chain/src/lib.rs]: Minor: if you're referencing a macro by path (like `static_assertions::macro!` directly or by having a `use static_assertions::macro;` in scope), then the `extern crate` statement is not necessary.

**Review by jvff — CHANGES_REQUESTED** [review]
I'm not sure if less than or less than equal to should be used in the assertion :thinking:

File: zebra-chain/src/lib.rs (hunk 4 above)
Comment (conradoplg) [reply zebra-chain/src/lib.rs]: Fixed in 1d4bf4a7, thanks!

You were also right about the `<`, I had even wrote like that at first but changed later, I got confused with `max_allocation()` being an inclusive bound

**Review by jvff — APPROVED** [review]

**Pull request merged**